Add Migration Guide for Opflow (#9549)
Qiskit/qiskit-terra: docs/migration_guides/opflow_migration.rst

=======================
Opflow Migration Guide
=======================

*Jump to* `TL;DR`_.

Background
----------

The :mod:`~qiskit.opflow` module was originally introduced as a layer between circuits and algorithms, a series of building blocks
for quantum algorithms research and development.

> The recent release of the :mod:`~qiskit.primitives` introduced a new interface for interacting with backends. Now, instead of

Reviewer: ```suggestion
The recent release of the :mod:`~qiskit.primitives` introduced a new paradigm for interacting with backends. Now, instead of
```

preparing a circuit to execute with a ``backend.run()`` type of method, the algorithms can leverage the :class:`~Sampler` and
:class:`~Estimator` primitives, send parametrized circuits and observables, and directly receive quasi-probability distributions or
expectation values (respectively). This workflow simplifies considerably the pre-processing and post-processing steps
that previously relied on opflow; encouraging us to move away from opflow, and find new paths of developing algorithms based on
the :mod:`~qiskit.primitives` interface and the :mod:`~qiskit.quantum_info` module.

This guide traverses the opflow submodules and provides either a direct alternative

> (i.e. using :mod:`~qiskit.quantum_info`), or an explanation of how to replace their functionality in algorithms.

Reviewer: I'm just going to comment here, before things gets into more detail. Not sure exactly where this might go but I am wondering whether as overview we can have a simple table, or just text that summarizes the function equivalency/replacement. E.g, if you used opflow operators and/or states look to quantum_info; for expectations look to the Estimator primitives. circuit sampler the Sampler primitives, gradients the new primitive based Gradients etc. This would give people an overview/idea upfront before jumping into detail/specifics about these

Author: I have added a table. The quantum info module's documentation is not separated in modules so the links are not ideal, but hopefully it transmits the message.

> TL;DR

Reviewer 2: I think the TL;DR section should be the first section, above the Background section

-----
.. note::

    The new :mod:`~qiskit.primitives`, in combination with the :mod:`~qiskit.quantum_info` module, have superseded
    functionality of :mod:`~qiskit.opflow`. Thus, the latter is being deprecated.

> Index

Reviewer 2: I think "Contents" works better than Index here

-----
This guide covers the migration from these opflow sub-modules:

**Operators**

- `Operator Base Class`_
- `Operator Globals`_
- `Primitive and List Ops`_
- `State Functions`_

**Converters**

- `Converters`_
- `Evolutions`_
- `Expectations`_

**Gradients**

- `Gradients`_

Operator Base Class
-------------------
*Back to* `Index`_

> The :class:`~opflow.OperatorBase` abstract class can generally be replaced with :class:`~quantum_info.BaseOperator`, keeping in

Reviewer: I had mentioned somewhere to one of the links, that my comment applied in general, but for instance opflow.operators are documented at that location, though the code does allow import at the opflow level, it was documented by module so they need to be in the links. To start with we might want the full path to show (i.e. without the leading ~) so its clear. Whether we could do any better to more readily distinguish OperatorBase and BaseOperator I don't know - given they are similar care is needed when reading!

> The :class:`~opflow.OperatorBase` abstract class can generally be replaced with :class:`~quantum_info.BaseOperator`, keeping in

Author: This is a good point, thanks for bringing it up! I added the full path to this section, and changed the ``tip`` box to an ``attention`` one. I think this already makes it clearer, but let me know if you want me to add more information.

mind that :class:`~quantum_info.BaseOperator` is more generic than its opflow counterpart.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative
   * - :class:`~opflow.OperatorBase`
     - :class:`~quantum_info.BaseOperator`

..  tip::

    Please note that:

    1. :class:`~opflow.OperatorBase` implements a broader algebra mixin. Some operator overloads are not available in
    :class:`~quantum_info.BaseOperator`.

    2. :class:`~opflow.OperatorBase` also implements methods such as ``.to_matrix()`` or ``.to_spmatrix()``, which are only found
    in some of the :class:`~quantum_info.BaseOperator` subclasses.

    See API reference for more information.

> Operator Globals

Reviewer: A general comment about headings etc. The above index/main headings - that seems to get lost when scrolling through the text as many headings seem to be text of a similar size making the structure when scrolling less than obvious. There are even examples with the same size font.
I guess it might be possible to break the text into different documents which would keep one within a given top level context too.

Author: I am not a big fan of breaking the text into different documents, but I have decided to change the format of the twisties, and that will hopefully also help to navigate the text better.

directly creating a corresponding :class:`~qiskit.quantum_info.SparsePauliOp`, as shown in the example :ref:`below <1_q_pauli>`.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative
   * - :class:`~qiskit.opflow.X`, :class:`~qiskit.opflow.Y`, :class:`~qiskit.opflow.Z`, :class:`~qiskit.opflow.I`
     - :class:`~qiskit.quantum_info.Pauli`

       ..  tip::

           For direct compatibility with classes in :mod:`~qiskit.algorithms`, wrap in :class:`~qiskit.quantum_info.SparsePauliOp`.

..  _1_q_pauli:

Example 1: Defining the XX operator
###################################
.. raw:: html

    <details>
    <summary><a><b>Opflow</b></a></summary>

.. code-block:: python

    from qiskit.opflow import X

    operator = X ^ X

.. raw:: html

   </details>

.. raw:: html

    <details>
    <summary><a><b>Alternative</b></a></summary>

.. code-block:: python

    from qiskit.quantum_info import Pauli, SparsePauliOp

    X = Pauli('X')
    op = X ^ X

    # equivalent to:
    op = SparsePauliOp('XX')

.. raw:: html

   </details>

Example 2: Defining a more complex operator
###########################################

> .. raw:: html

Reviewer: In regards of the twisties. I find the information behind them the stuff I want to read and I want to see both so having them on independent twisties I personally find awkward. I guess I find it easier to see it and be quickly able to scroll by if it does not appear what I want rather that having to keep going for twisties. I could see them for extra very similar examples perhaps. Maybe others have a different experience with them than I...

Author: I see what you mean. I have added 1 twistie per example, instead of per alternative, I think that makes more sense.

    <details>
    <summary><a><b>Opflow</b></a></summary>

.. code-block:: python

    from qiskit.opflow import I, X, Z

    op = 0.39 * (I ^ Z ^ I) + 0.5 * (I ^ X ^ X)

.. raw:: html

   </details>

.. raw:: html

    <details>
    <summary><a><b>Alternative</b></a></summary>

.. code-block:: python

    from qiskit.quantum_info import SparsePauliOp

    op = SparsePauliOp.from_list([("IZI", 0.39), ("IXX", 0.5)])

    # or...
    op = SparsePauliOp.from_sparse_list([("Z", [1], 0.39), ("XX", [0,1], 0.5)], num_qubits = 3)

> 

Reviewer: One can also directly build it using its constructor
```
op = SparsePauliOp(["IZI", "IXX"], coeffs=[0.39, 0.5])
```

Author: added

It also acts as a factory to instantiate a corresponding sub-class depending on the computational primitive used
to initialize it.

.. tip::

    Interpreting :class:`~qiskit.opflow.primitive_ops.PrimitiveOp` as a factory class:

    .. list-table::
       :header-rows: 1

       * - Class passed to :class:`~qiskit.opflow.primitive_ops.PrimitiveOp`
         - Sub-class returned

       * - :class:`~qiskit.quantum_info.Pauli`
         - :class:`~qiskit.opflow.primitive_ops.PauliOp`

       * - :class:`~qiskit.circuit.Instruction`, :class:`~qiskit.circuit.QuantumCircuit`
         - :class:`~qiskit.opflow.primitive_ops.CircuitOp`

       * - ``list``, ``np.ndarray``, ``scipy.sparse.spmatrix``, :class:`~qiskit.quantum_info.Operator`
         - :class:`~qiskit.opflow.primitive_ops.MatrixOp`

Thus, when migrating opflow code, it is important to look for alternatives to replace the specific subclasses that
might have been used "under the hood" in the original code:

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative

   * - :class:`~qiskit.opflow.primitive_ops.PrimitiveOp`

>      - :class:`~qiskit.quantum_info.Operator`

Reviewer 3: As you explained nicely above, the `PrimitiveOp` was more a factory and base class, so I wouldn't say that `Operator` is a replacement as that strictly stores matrix representations 🙂 I think this just doesn't have (or need) an alternative

   * - :class:`~qiskit.opflow.primitive_ops.CircuitOp`

>      - :class:`~qiskit.quantum_info.Operator`

Reviewer 3: This should be `QuantumCircuit` right?
```suggestion
     - :class:`~qiskit.QuantumCircuit`
```

    # can be represented as:
    tapered_op = TaperedPauliSumOp(primitive, z2_symmetries)

.. raw:: html

   </details>

.. raw:: html

    <details>
    <summary><a><b>Alternative</b></a></summary>

.. code-block:: python

    from qiskit.quantum_info import SparsePauliOp, Z2Symmetries

    qubit_op = SparsePauliOp.from_list(
        [
            ("II", -1.0537076071291125),
            ("IZ", 0.393983679438514),
            ("ZI", -0.39398367943851387),
            ("ZZ", -0.01123658523318205),
            ("XX", 0.1812888082114961),
        ]
    )
    z2_symmetries = Z2Symmetries.find_z2_symmetries(qubit_op)
    tapered_op = z2_symmetries.taper(qubit_op)

.. raw:: html

   </details>

ListOps
~~~~~~~
*Back to* `Index`_

The :mod:`~qiskit.opflow.list_ops` module contained classes for manipulating lists of :mod:`~qiskit.opflow.primitive_ops`
or :mod:`~qiskit.opflow.state_fns`. The :mod:`~qiskit.quantum_info` alternatives for this functionality are the
:class:`~qiskit.quantum_info.PauliList`, :class:`~qiskit.quantum_info.SparsePauliOp` (for sums of ``Pauli``\s).

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative

   * - :class:`~qiskit.opflow.list_ops.ListOp`

>      - No direct replacement. This is the base class for operator lists. In general, these could be replaced with

Reviewer 2: a lot of these have no direct replacement, is it worth adding a sentence to the paragraph above about why no direct replacement is needed? Or why the choice was made not to replace them?

       * - Class passed to :class:`~qiskit.opflow.state_fns.StateFn`
         - Sub-class returned

       * - ``str``, ``dict``, :class:`~qiskit.result.Result`
         - :class:`~qiskit.opflow.state_fns.DictStateFn`

       * - ``list``, ``np.ndarray``, :class:`~qiskit.quantum_info.Statevector`
         - :class:`~qiskit.opflow.state_fns.VectorStateFn`

       * - :class:`~qiskit.circuit.QuantumCircuit`, :class:`~qiskit.circuit.Instruction`
         - :class:`~qiskit.opflow.state_fns.CircuitStateFn`

       * - :class:`~qiskit.opflow.OperatorBase`
         - :class:`~qiskit.opflow.state_fns.OperatorStateFn`

This means that references to :class:`~qiskit.opflow.state_fns.StateFn` in opflow code should be examined to
identify the sub-class that is being used, to then look for an alternative.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative

   * - :class:`~qiskit.opflow.state_fns.StateFn`
     - In most cases, :class:`~qiskit.quantum_info.Statevector`. Remember that this is a factory class.

   * - :class:`~qiskit.opflow.state_fns.CircuitStateFn`
     - :class:`~qiskit.quantum_info.Statevector`

   * - :class:`~qiskit.opflow.state_fns.DictStateFn`

>      - :class:`~qiskit.quantum_info.Statevector`

Reviewer 3: The ``DictStateFn`` could store an efficient representation if the measurement results were sparse, which the ``Statevector`` cannot, so I think a more accurate replacement would be the ``qiskit.result.QuasiDist`` or the ``Result`` class you mentioned above 🙂

Author: Yes, I forgot to update this part of the table, thanks for catching it!

   * - :class:`~qiskit.opflow.state_fns.VectorStateFn`
     - This class can be replaced with :class:`~qiskit.quantum_info.Statevector` or
       :class:`~qiskit.quantum_info.StabilizerState` (for Clifford-based vectors).

   * - :class:`~qiskit.opflow.state_fns.SparseVectorStateFn`
     - No direct replacement.

   * - :class:`~qiskit.opflow.state_fns.OperatorStateFn`

>      - :class:`~qiskit.quantum_info.Statevector`

Reviewer 3: This also doesn't have a replacement, the ``OperatorStateFn`` was to represent measurements again operators 🙂

   * - :class:`~qiskit.opflow.state_fns.CVaRMeasurement`
     - Used in :class:`~qiskit.opflow.expectations.CVaRExpectation`.

>        Functionality now covered by :class:`~SamplingEstimator`. See :ref:`example <cvar>` in expectations.

Reviewer 3: ```suggestion
       Functionality now covered by :class:`.SamplingVQE`. See :ref:`example <cvar>` in expectations.
```

Example 1: Applying an operator to a state
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. raw:: html

    <details>
    <summary><a><b>Opflow</b></a></summary>

.. code-block:: python

    from qiskit.opflow import StateFn, X, Y

    qc = QuantumCircuit(2)
    op = X ^ Y
    state = StateFn(qc)

>     comp = ~state @ op

Reviewer 3: This implements `<state | op` -- would `op | state>` be more intuitive to use? 😄 The latter is also what `Statevector.evolve` implements, I believe.

Author: Yes, I will change it.

    # returns a VectorStateFn (Statevector)

.. raw:: html

   </details>

.. raw:: html

    <details>
    <summary><a><b>Alternative</b></a></summary>

.. code-block:: python

    from qiskit import QuantumCircuit
    from qiskit.quantum_info import SparsePauliOp, Statevector

    qc = QuantumCircuit(2)
    op = SparsePauliOp("XY")
    state = Statevector(qc)

    eval = state.evolve(operator)
    # returns a Statevector

.. raw:: html

   </details>

See more applied examples in :ref:`expectations <expect_state>`  and :ref:`converters <convert_state>`.

Converters
----------
*Back to* `Index`_

The role of this sub-module was to convert the operators into other opflow operator classes
(:class:`~qiskit.opflow.converters.TwoQubitReduction`, :class:`~qiskit.opflow.converters.PauliBasisChange`...).
In the case of the :class:`~qiskit.opflow.converters.CircuitSampler`, it traversed an operator and outputted
approximations of its state functions using a quantum backend.
Notably, this functionality has been replaced by the :mod:`~qiskit.primitives`.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative

   * - :class:`~qiskit.opflow.converters.CircuitSampler`

>      - :class:`~primitives.Estimator`

Reviewer: I thought the CircuitSampler was more akin to Sampler. It takes States (described by circuits) and executes them to into Dict or Vector states akin to the probabilities. There are the expectations and basis change logic that used in conjunction with this ends up being more akin to Estimator.

Reviewer 3: As the circuit sampler was the base for all circuit evaluations, I'd think that both Sampler and Estimator are a replacement for this.

>      - :class:`~primitives.Estimator`

Reviewer 3: ```suggestion
     - :class:`~qiskit.primitives.Estimator`
```

.. raw:: html

    <details>
    <summary><a><b>Opflow</b></a></summary>

.. code-block:: python

    from qiskit_aer import Aer
    from qiskit.circuit import QuantumCircuit, Parameter
    from qiskit.opflow import ListOp, StateFn, CircuitSampler

    x, y = Parameter("x"), Parameter("y")

    circuit1 = QuantumCircuit(1)
    circuit1.p(0.2, 0)
    circuit2 = QuantumCircuit(1)
    circuit2.p(x, 0)
    circuit3 = QuantumCircuit(1)
    circuit3.p(y, 0)

    bindings = {x: -0.4, y: 0.4}
    listop = ListOp([StateFn(circuit) for circuit in [circuit1, circuit2, circuit3]])

    sampler = CircuitSampler(Aer.get_backend("aer_simulator"))
    sampled = sampler.convert(listop, params=bindings).eval()
    # returns list of SparseVectorStateFn

.. raw:: html

    </details>

.. raw:: html

    <details>
    <summary><a><b>Alternative</b></a></summary>

.. code-block:: python

    from qiskit.circuit import QuantumCircuit, Parameter
    from qiskit.primitives import Sampler

    x, y = Parameter("x"), Parameter("y")

    circuit1 = QuantumCircuit(1)
    circuit1.p(0.2, 0)
    circuit2 = QuantumCircuit(1)
    circuit2.p(x, 0)
    circuit3 = QuantumCircuit(1)
    circuit3.p(y, 0)

>     circuits = [circuit1, circuit2, circuit3]

Reviewer 3: The circuits miss measurements 🙂

Author: He :sweat_smile: . I think I will have to make sure to run all of the examples in case there are more of these laying around.

>     param_values = [[-0.4, 0.4] for _ in circuits]

Reviewer 3: Would it be better to do only give the actual parameter values for each circuit?
```suggestion
    param_values = [None, [-0.4], [0.4]]
```

.. raw:: html

    </details>

Example 2: ``CircuitSampler`` for computing Expectation Values
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. raw:: html

    <details>
    <summary><a><b>Opflow</b></a></summary>

.. code-block:: python

    from qiskit import QuantumCircuit
    from qiskit.opflow import X, Z, StateFn, CircuitStateFn, CircuitSampler
    from qiskit.providers.aer import AerSimulator

    qc = QuantumCircuit(1)
    qc.h(0)
    state = CircuitStateFn(qc)
    hamiltonian = X + Z

    expr = StateFn(hamiltonian, is_measurement=True).compose(state)
    backend = AerSimulator()
    sampler = CircuitSampler(backend)
    expectation = sampler.convert(expr)
    expectation_value = expectation.eval().real

.. raw:: html

    </details>

.. raw:: html

    <details>
    <summary><a><b>Alternative</b></a></summary>

.. code-block:: python

    from qiskit import QuantumCircuit
    from qiskit.primitives import Estimator
    from qiskit.quantum_info import SparsePauliOp

    state = QuantumCircuit(1)
    state.h(0)
    hamiltonian = SparsePauliOp.from_list([('X', 1), ('Z',1)])

    estimator = Estimator()

>     expectation_value = estimator.run(state, hamiltonian).result().values

Reviewer 3: Very nitpicky, but to be equal to the above do we need `.real` here too?
```suggestion
    expectation_value = estimator.run(state, hamiltonian).result().values.real
```

.. raw:: html

    </details>

Evolutions
----------
*Back to* `Index`_

> The :mod:`~qiskit.opflow.evolutions` sub-module was created to provide building blocks for hamiltonian simulation algorithms,

Reviewer 3: I'm on a nitpicky streak here, but I think since `Hamiltonian` is a name it's usually spelt in uppercase 🙂 
```suggestion
The :mod:`~qiskit.opflow.evolutions` sub-module was created to provide building blocks for Hamiltonian simulation algorithms,
```

    perform the trotterization (either via a ``string``\, which is then inputted into a :class:`~qiskit.opflow.evolutions.TrotterizationFactory`,
    or by supplying a method instance of :class:`~qiskit.opflow.evolutions.Trotter`, :class:`~qiskit.opflow.evolutions.Suzuki` or :class:`~qiskit.opflow.evolutions.QDrift`).

    The different trotterization methods that extend :class:`~qiskit.opflow.evolutions.TrotterizationBase` were migrated to :mod:`~qiskit.synthesis`,
    and now extend the :class:`~qiskit.synthesis.evolution.ProductFormula` base class. They no longer contain a ``.convert()`` method for
    standalone use, but now are designed to be plugged into the :class:`~qiskit.synthesis.PauliEvolutionGate` and called via ``.synthesize()``.
    In this context, the job of the :class:`~qiskit.opflow.evolutions.PauliTrotterEvolution` class can now be handled directly by the algorithms
    (for example, :class:`~qiskit.algorithms.time_evolvers.TrotterQRTE`\).

    In a similar manner, the :class:`~qiskit.opflow.evolutions.MatrixEvolution` class performs evolution by classical matrix exponentiation,

>     constructing a circuit with :class:`~UnitaryGate`\s or :class:`~HamiltonianGate`\s containing the exponentiation of the operator.
>     This class is no longer necessary, as the :class:`~HamiltonianGate`\s can be directly handled by the algorithms.

Reviewer 3: ```suggestion
    constructing a circuit with :class:`.UnitaryGate`\s or :class:`.HamiltonianGate`\s containing the exponentiation of the operator.
    This class is no longer necessary, as the :class:`.HamiltonianGate`\s can be directly handled by the algorithms.
```

   * - :class:`~qiskit.opflow.evolutions.Suzuki`
     - :class:`~synthesis.SuzukiTrotter`

   * - :class:`~qiskit.opflow.evolutions.QDrift`
     - :class:`~synthesis.QDrift`

Other Evolution Classes
~~~~~~~~~~~~~~~~~~~~~~~~
*Back to* `Index`_

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative

   * - :class:`~qiskit.opflow.evolutions.EvolutionFactory`
     - No direct replacement. This class was used to create instances of one of the classes listed below.

   * - :class:`~qiskit.opflow.evolutions.EvolvedOp`
     - No direct replacement. The workflow no longer requires a specific operator for evolutions.

   * - :class:`~qiskit.opflow.evolutions.MatrixEvolution`
     - :class:`~HamiltonianGate`

   * - :class:`~qiskit.opflow.evolutions.PauliTrotterEvolution`
     - :class:`~PauliEvolutionGate`

Example 1: Trotter evolution
############################

.. raw:: html

    <details>
    <summary><a><b>Opflow</b></a></summary>

.. code-block:: python

    from qiskit.opflow import Trotter, PauliTrotterEvolution, PauliSumOp

    hamiltonian = PauliSumOp.from_list([('X', 1), ('Z',1)])
    evolution = PauliTrotterEvolution(trotter_mode=Trotter(), reps=1)

>     evol_result = evolution.convert(hamiltonian.exp_i())

Reviewer 3: I think it would be cool to show how the `time` parameter is handled:
```
(time * hamiltonian).exp_i()
```
vs
```
PauliEvolutionGate(hamiltonian, time=time)
```
😄

    evolved_state = evol_result.to_circuit()
.. raw:: html

    </details>

.. raw:: html

    <details>
    <summary><a><b>Alternative</b></a></summary>

.. code-block:: python

            from qiskit.quantum_info import SparsePauliOp
            from qiskit.synthesis import SuzukiTrotter
            from qiskit.circuit.library import PauliEvolutionGate
            from qiskit import QuantumCircuit

            hamiltonian = SparsePauliOp.from_list([('X', 1), ('Z',1)])

>             evol_gate = PauliEvolutionGate(hamiltonian, 1, synthesis=SuzukiTrotter())

Reviewer 3: I would expect the replacement to be `LieTrotter` here, which implements `reps=1` -- but maybe I'm wrong!

    evolved_state.append(evol_gate, [0])

.. raw:: html

    </details>

Expectations
------------
*Back to* `Index`_

Expectations are converters which enable the computation of the expectation value of an observable with respect to some state function.
This functionality can now be found in the estimator primitive.

Algorithm-Agnostic Expectations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*Back to* `Index`_

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative

   * - :class:`~qiskit.opflow.expectations.ExpectationFactory`
     - No direct replacement. This class was used to create instances of one of the classes listed below.

   * - :class:`~qiskit.opflow.expectations.AerPauliExpectation`
     - Use :class:`~Estimator` primitive from ``qiskit_aer`` with ``approximation=True`` and ``shots=None`` as ``run_options``.
       See example :ref:`below <expect_state>`.

   * - :class:`~qiskit.opflow.expectations.MatrixExpectation`
     - Use :class:`~Estimator` primitive from ``qiskit`` (if no shots are set, it performs an exact Statevector calculation).
       See example :ref:`below <matrix_state>`.

   * - :class:`~qiskit.opflow.expectations.PauliExpectation`

>      - Use any :class:`~Estimator` primitive from ``qiskit``.

Reviewer 3: Here we have to be a bit careful and mention that shots have to be set or that users should use the Aer estimator 🙂

> =======================
> Opflow Migration Guide
> =======================

Reviewer 4: ```suggestion
#######################
Opflow Migration Guide
#######################
```

> TL;DR
> -----

Reviewer 4: ```suggestion
TL;DR
=====
```

The new :mod:`~qiskit.primitives`, in combination with the :mod:`~qiskit.quantum_info` module, have superseded
functionality of :mod:`~qiskit.opflow`. Thus, the latter is being deprecated.

.. note::

    The use of :mod:`~qiskit.opflow` was tightly coupled to the :class:`~qiskit.utils.QuantumInstance` class, which
    is also being deprecated. For more information on migrating the :class:`~qiskit.utils.QuantumInstance`, please

>     read the `quantum instance migration guide <http://qisk.it/qi_migration>`_.

Author: Yes, that is the idea with these shortlinks, we can redirect them to the published docs once they are merged.

>     read the `quantum instance migration guide <http://qisk.it/qi_migration>`_.

Author: Well, right now there is nothing to link to, but once these are merged and the docs exist, I would have no problem updating this and adding the proper reference :)

> Background
> ----------

Reviewer 4: ```suggestion
Background
==========
```

:class:`.Estimator` primitives, send parametrized circuits and observables, and directly receive quasi-probability distributions or
expectation values (respectively). This workflow simplifies considerably the pre-processing and post-processing steps
that previously relied on this module; encouraging us to move away from :mod:`~qiskit.opflow`
and find new paths for developing algorithms based on the :mod:`~qiskit.primitives` interface and
the :mod:`~qiskit.quantum_info` module.

This guide traverses the opflow submodules and provides either a direct alternative
(i.e. using :mod:`~qiskit.quantum_info`), or an explanation of how to replace their functionality in algorithms.

The function equivalency can be roughly summarized as follows:

.. |operator_globals| replace:: ``operator_globals``
.. _operator_globals: https://qiskit.org/documentation/apidoc/opflow.html#operator-globals/

.. list-table::
   :header-rows: 1

   * - Opflow Module
     - Alternative
   * - Operators (:class:`~qiskit.opflow.OperatorBase`, |operator_globals|_ ,
       :mod:`~qiskit.opflow.primitive_ops`,
       :mod:`~qiskit.opflow.list_ops`\)

>      - :mod:`qiskit.quantum_info` *Operators*

Reviewer 4: If you add a label before the section you want to refer to, you can link to the section directly, see: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-arbitrary-locations

In this case, you can add `.. _quantum_info_operators:` in front of the `Operators` section of the `quantum_info` module level API.

.. _qiskit_aer.primitives: https://qiskit.org/documentation/locale/de_DE/apidoc/aer_primitives.html

.. |qiskit_aer.primitives.Estimator| replace:: ``qiskit_aer.primitives.Estimator``
.. _qiskit_aer.primitives.Estimator: https://qiskit.org/documentation/locale/de_DE/stubs/qiskit_aer.primitives.Estimator.html

.. |qiskit_ibm_runtime| replace:: ``qiskit_ibm_runtime``
.. _qiskit_ibm_runtime: https://qiskit.org/documentation/partners/qiskit_ibm_runtime/primitives.html

..  attention::

    Most references to the :class:`qiskit.primitives.Sampler` or :class:`qiskit.primitives.Estimator` in this guide
    can be replaced with instances of the Aer primitives (|qiskit_aer.primitives|_ ), Runtime primitives
    (|qiskit_ibm_runtime|_ ) or Terra backend primitives (:class:`qiskit.primitives.BackendSampler`,
    :class:`qiskit.primitives.BackendEstimator`). Certain classes, such as the
    :class:`~qiskit.opflow.expectations.AerPauliExpectation`, are only replaced by a specific primitive instance
    (in this case, |qiskit_aer.primitives.Estimator|_ ), or require a specific option configuration.
    This will be explicitly indicated in the corresponding section.

> Contents
> --------

Reviewer 4: ```suggestion
Contents
========
```

This document covers the migration from these opflow sub-modules:

**Operators**

- `Operator Base Class`_
- `Operator Globals`_
- `Primitive and List Ops`_
- `State Functions`_

**Converters**

- `Converters`_
- `Evolutions`_
- `Expectations`_

**Gradients**

- `Gradients`_

> Operator Base Class
> -------------------

Reviewer 4: ```suggestion
Operator Base Class
===================
```

keeping in mind that |qiskit.quantum_info.BaseOperator|_ is more generic than its opflow counterpart.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative
   * - :class:`qiskit.opflow.OperatorBase`
     - |qiskit.quantum_info.BaseOperator|_

..  attention::

    Despite the similar class names, :class:`qiskit.opflow.OperatorBase` and
    |qiskit.quantum_info.BaseOperator|_ are not completely equivalent to each other, and the transition
    should be handled with care. Namely:

    1. :class:`qiskit.opflow.OperatorBase` implements a broader algebra mixin. Some operator overloads that were
    commonly used :mod:`~qiskit.opflow` (for example ``~`` for ``.adjoint()``) are not defined for
    |qiskit.quantum_info.BaseOperator|_. You might want to check the specific
    :mod:`~qiskit.quantum_info` subclass instead.

    2. :class:`qiskit.opflow.OperatorBase` also implements methods such as ``.to_matrix()`` or ``.to_spmatrix()``,
    which are only found in some of the |qiskit.quantum_info.BaseOperator|_ subclasses.

    See API reference for more information.

> Operator Globals
> ----------------

Reviewer 4: ```suggestion
Operator Globals
================
```

*Back to* `Contents`_

Opflow provided shortcuts to define common single qubit states, operators, and non-parametrized gates in the
|operator_globals|_ module.

These were mainly used for didactic purposes or quick prototyping, and can easily be replaced by their corresponding
:mod:`~qiskit.quantum_info` class: :class:`~qiskit.quantum_info.Pauli`, :class:`~qiskit.quantum_info.Clifford` or
:class:`~qiskit.quantum_info.Statevector`.

> 1-Qubit Paulis
> ~~~~~~~~~~~~~~

Reviewer 4: ```suggestion
1-Qubit Paulis
~~~~~~~~~~~~~~
--------------
```

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 1: Defining the XX operator</font></a></summary>
    <br>

**Opflow**

.. code-block:: python

    from qiskit.opflow import X

    operator = X ^ X

**Alternative**

.. code-block:: python

    from qiskit.quantum_info import Pauli, SparsePauliOp

    X = Pauli('X')
    op = X ^ X

    # equivalent to:
    op = Pauli('XX')

    # equivalent to:
    op = SparsePauliOp('XX')

.. raw:: html

   </details>

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 2: Defining a more complex operator</font></a></summary>
    <br>

**Opflow**

.. code-block:: python

    from qiskit.opflow import I, X, Z, PauliSumOp

    op = 0.39 * (I ^ Z ^ I) + 0.5 * (I ^ X ^ X)

    # or ...
    op = PauliSumOp.from_list([("IZI", 0.39), ("IXX", 0.5)])

**Alternative**

.. code-block:: python

    from qiskit.quantum_info import SparsePauliOp

    op = SparsePauliOp(["IZI", "IXX"], coeffs = [0.39, 0.5])

    # or...
    op = SparsePauliOp.from_list([("IZI", 0.39), ("IXX", 0.5)])

    # or...
    op = SparsePauliOp.from_sparse_list([("Z", [1], 0.39), ("XX", [0,1], 0.5)], num_qubits = 3)

.. raw:: html

   </details>

> Common non-parametrized gates (Clifford)
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 4: ```suggestion
Common non-parametrized gates (Clifford)
----------------------------------------
```

       :class:`~qiskit.quantum_info.Operator`\s can be also directly constructed from quantum circuits.
       Another alternative is to wrap the circuit in :class:`~qiskit.quantum_info.Clifford` and call
       ``Clifford.to_operator()``.

       ..  note::

            Constructing :mod:`~qiskit.quantum_info` operators from circuits is not efficient, as it is a dense operation and
            scales exponentially with the size of the circuit, use with care.

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 1: Defining the HH operator</font></a></summary>
    <br>

**Opflow**

.. code-block:: python

    from qiskit.opflow import H

    op = H ^ H

**Alternative**

.. code-block:: python

    from qiskit import QuantumCircuit
    from qiskit.quantum_info import Clifford, Operator

    qc = QuantumCircuit(2)
    qc.h(0)
    qc.h(1)
    op = Clifford(qc).to_operator()

    # or...
    qc = QuantumCircuit(1)
    qc.h(0)
    H = Clifford(qc).to_operator()
    op = H ^ H

    # or, directly
    qc = QuantumCircuit(2)
    qc.h(0)
    qc.h(1)
    op = Operator(qc)

.. raw:: html

   </details>

> 1-Qubit States
> ~~~~~~~~~~~~~~

Reviewer 4: ```suggestion
1-Qubit States
--------------
```

     - :class:`~qiskit.quantum_info.Statevector` or simply :class:`~qiskit.circuit.QuantumCircuit`, depending on the use case.

       ..  note::

           For efficient simulation of stabilizer states, :mod:`~qiskit.quantum_info` includes a
           :class:`~qiskit.quantum_info.StabilizerState` class. See API ref. for more info.

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 1: Working with stabilizer states</font></a></summary>
    <br>

**Opflow**

.. code-block:: python

    from qiskit.opflow import Zero, One, Plus, Minus

    # Zero, One, Plus, Minus are all stabilizer states
    state1 = Zero ^ One
    state2 = Plus ^ Minus

**Alternative**

.. code-block:: python

    from qiskit import QuantumCircuit
    from qiskit.quantum_info import StabilizerState, Statevector

    qc_zero = QuantumCircuit(1)
    qc_one = qc_zero.copy()
    qc_one.x(0)
    state1 = Statevector(qc_zero) ^ Statevector(qc_one)

    qc_plus = qc_zero.copy()
    qc_plus.h(0)
    qc_minus = qc_one.copy()
    qc_minus.h(0)

    state2 = StabilizerState(qc_plus) ^ StabilizerState(qc_minus)

.. raw:: html

   </details>

> Primitive and List Ops
> ----------------------

Reviewer 4: ```suggestion
Primitive and List Ops
======================
```

*Back to* `Contents`_

Most of the workflows that previously relied in components from :mod:`~qiskit.opflow.primitive_ops` and
:mod:`~qiskit.opflow.list_ops` can now leverage elements from :mod:`~qiskit.quantum_info`\'s
operators instead.
Some of these classes do not require a 1-1 replacement because they were created to interface with other
opflow components.

Primitive Ops

> ~~~~~~~~~~~~~~

Reviewer 4: Can you change the remaining sections and subsections to follow this convention?

```
##########
Title name
##########

Section name
============

Subsection name
---------------

Subsubsection name
^^^^^^^^^^^^^^^^^^

Paragraph name
""""""""""""""
```

Author: Done!

Author: I will go ahead and change these in my other migration guides, now that I know this convention exists.

Reviewer 4: Thanks! I will focus on reviewing this PR for the moment. Once we merge this one we can move onto the other two migration guides. But it's good that you can already apply changes suggested for this migration for the others.

#######################
Opflow Migration Guide
#######################

TL;DR
=====
The new :mod:`~qiskit.primitives`, in combination with the :mod:`~qiskit.quantum_info` module, have superseded
functionality of :mod:`~qiskit.opflow`. Thus, the latter is being deprecated.

In this migration guide, you will find instructions and code examples for how to migrate your code based on
the :mod:`~qiskit.opflow` module to the :mod:`~qiskit.primitives` and :mod:`~qiskit.quantum_info` modules.

.. note::

    The use of :mod:`~qiskit.opflow` was tightly coupled to the :class:`~qiskit.utils.QuantumInstance` class, which
    is also being deprecated. For more information on migrating the :class:`~qiskit.utils.QuantumInstance`, please
    read the `quantum instance migration guide <http://qisk.it/qi_migration>`_.

.. _attention_primitives:

..  attention::

    Most references to the :class:`qiskit.primitives.Sampler` or :class:`qiskit.primitives.Estimator` in this guide
    can be replaced with instances of the:

>     - Aer primitives (:class:`qiskit_aer.primitives.Sampler`, :class:`qiskit_aer.primitives.Estimator`)

Reviewer 4: As mentioned in the comment in `conf.py`, you will need to add `qiskit_aer` to intersphinx mapping for these links to be generated.

Author: Counting on https://github.com/Qiskit/qiskit-terra/pull/9716 to do that

and find new paths for developing algorithms based on the :mod:`~qiskit.primitives` interface and
the :mod:`~qiskit.quantum_info` module.

This guide traverses the opflow submodules and provides either a direct alternative
(i.e. using :mod:`~qiskit.quantum_info`), or an explanation of how to replace their functionality in algorithms.

The function equivalency can be roughly summarized as follows:

.. list-table::
   :header-rows: 1

   * - Opflow Module
     - Alternative
   * - Operators (:class:`~qiskit.opflow.OperatorBase`, :ref:`operator_globals`,
       :mod:`~qiskit.opflow.primitive_ops`, :mod:`~qiskit.opflow.list_ops`)
     - ``qiskit.quantum_info`` :ref:`Operators <quantum_info_operators>`

   * - :mod:`qiskit.opflow.state_fns`
     - ``qiskit.quantum_info`` :ref:`States <quantum_info_states>`

   * - :mod:`qiskit.opflow.converters`
     - :mod:`qiskit.primitives`

   * - :mod:`qiskit.opflow.evolutions`
     - ``qiskit.synthesis`` :ref:`Evolution <evolution_synthesis>`

   * - :mod:`qiskit.opflow.expectations`
     - :class:`qiskit.primitives.Estimator`

   * - :mod:`qiskit.opflow.gradients`
     - :mod:`qiskit.algorithms.gradients`

Contents
========

> This document covers the migration from these opflow sub-modules:

Reviewer 5: ```suggestion
This document covers the migration from these opflow submodules:
```
For consistency.

        state1 = Zero ^ One
        state2 = Plus ^ Minus

        print("State 1: ", state1)
        print("State 2: ", state2)

    .. testoutput::

        State 1:  DictStateFn({'01': 1})
        State 2:  CircuitStateFn(
             ┌───┐┌───┐
        q_0: ┤ X ├┤ H ├
             ├───┤└───┘
        q_1: ┤ H ├─────
             └───┘
        )

    **Alternative**

    .. testcode::

        from qiskit import QuantumCircuit
        from qiskit.quantum_info import StabilizerState, Statevector

        qc_zero = QuantumCircuit(1)
        qc_one = qc_zero.copy()
        qc_one.x(0)
        state1 = Statevector(qc_zero) ^ Statevector(qc_one)
        print("State 1: ", state1)

        qc_plus = qc_zero.copy()
        qc_plus.h(0)
        qc_minus = qc_one.copy()
        qc_minus.h(0)
        state2 = StabilizerState(qc_plus) ^ StabilizerState(qc_minus)
        print("State 2: ", state2)

    .. testoutput::

        State 1:  Statevector([0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
                    dims=(2, 2))
        State 2:  StabilizerState(StabilizerTable: ['-IX', '+XI'])

Primitive and List Ops
======================
*Back to* `Contents`_

> Most of the workflows that previously relied in components from :mod:`~qiskit.opflow.primitive_ops` and

Reviewer 5: ```suggestion
Most of the workflows that previously relied on components from :mod:`~qiskit.opflow.primitive_ops` and
```

     - No direct replacement. This is the base class for operator lists. In general, these could be replaced with
       Python ``list``\s. For :class:`~qiskit.quantum_info.Pauli` operators, there are a few alternatives, depending on the use-case.
       One alternative is :class:`~qiskit.quantum_info.PauliList`.

   * - :class:`~qiskit.opflow.list_ops.ComposedOp`
     - No direct replacement. Current workflows do not require composition of states and operators within
       one object (no lazy evaluation).

   * - :class:`~qiskit.opflow.list_ops.SummedOp`
     - No direct replacement. For :class:`~qiskit.quantum_info.Pauli` operators, use :class:`~qiskit.quantum_info.SparsePauliOp`.

   * - :class:`~qiskit.opflow.list_ops.TensoredOp`
     - No direct replacement. For :class:`~qiskit.quantum_info.Pauli` operators, use :class:`~qiskit.quantum_info.SparsePauliOp`.

State Functions
===============
*Back to* `Contents`_

The :mod:`~qiskit.opflow.state_fns` module can be generally replaced by subclasses of :mod:`~qiskit.quantum_info`\'s
:class:`qiskit.quantum_info.QuantumState`.

Similarly to :class:`~qiskit.opflow.primitive_ops.PrimitiveOp`, :class:`~qiskit.opflow.state_fns.StateFn`

> acts as a factory to create the corresponding sub-class depending on the computational primitive used to initialize it.

Reviewer 5: ```suggestion
acts as a factory to create the corresponding subclass depending on the computational primitive used to initialize it.
```
For consistency.

.. tip::

    Interpreting :class:`~qiskit.opflow.state_fns.StateFn` as a factory class:

    .. list-table::
       :header-rows: 1

       * - Class passed to :class:`~qiskit.opflow.state_fns.StateFn`
         - Sub-class returned

       * - ``str``, ``dict``, :class:`~qiskit.result.Result`
         - :class:`~qiskit.opflow.state_fns.DictStateFn`

       * - ``list``, ``np.ndarray``, :class:`~qiskit.quantum_info.Statevector`
         - :class:`~qiskit.opflow.state_fns.VectorStateFn`

       * - :class:`~qiskit.circuit.QuantumCircuit`, :class:`~qiskit.circuit.Instruction`
         - :class:`~qiskit.opflow.state_fns.CircuitStateFn`

       * - :class:`~qiskit.opflow.OperatorBase`
         - :class:`~qiskit.opflow.state_fns.OperatorStateFn`

This means that references to :class:`~qiskit.opflow.state_fns.StateFn` in opflow code should be examined to

> identify the sub-class that is being used, to then look for an alternative.

Reviewer 5: ```suggestion
identify the subclass that is being used, to then look for an alternative.
```

        print(repr(eval))

    .. testoutput::

        CircuitStateFn(
             ┌───┐
        q_0: ┤ X ├
             ├───┤
        q_1: ┤ Z ├
             └───┘
        )
        CircuitStateFn(
             ┌───┐┌────────────┐
        q_0: ┤ X ├┤0           ├
             ├───┤│  Pauli(XY) │
        q_1: ┤ Z ├┤1           ├
             └───┘└────────────┘
        )
        VectorStateFn(Statevector([ 0.0e+00+0.j,  0.0e+00+0.j, -6.1e-17-1.j,  0.0e+00+0.j],
                    dims=(2, 2)), coeff=1.0, is_measurement=False)

    **Alternative**

    .. testcode::

        from qiskit import QuantumCircuit
        from qiskit.quantum_info import SparsePauliOp, Statevector

        qc = QuantumCircuit(2)
        qc.x(0)
        qc.z(1)
        op = SparsePauliOp("XY")
        state = Statevector(qc)

        eval = state.evolve(op)

        print(state)
        print(eval)

    .. testoutput::

        Statevector([0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
                    dims=(2, 2))
        Statevector([0.+0.j, 0.+0.j, 0.-1.j, 0.+0.j],
                    dims=(2, 2))

See more applied examples in `Expectations`_  and `Converters`_.

Converters
==========

*Back to* `Contents`_

> The role of the :class:`qiskit.opflow.converters` sub-module was to convert the operators into other opflow operator classes

Reviewer 5: ```suggestion
The role of the :class:`qiskit.opflow.converters` submodule was to convert the operators into other opflow operator classes
```
For consistency.

        print(repr(grouped_sum))

    .. testoutput::

        SummedOp([PauliSumOp(SparsePauliOp(['XX'],
                      coeffs=[2.+0.j]), coeff=1.0), PauliSumOp(SparsePauliOp(['YY'],
                      coeffs=[1.+0.j]), coeff=1.0), PauliSumOp(SparsePauliOp(['IZ', 'ZZ'],
                      coeffs=[0.+2.j, 0.+1.j]), coeff=1.0)], coeff=1.0, abelian=False)

    **Alternative**

    .. testcode::

        from qiskit.quantum_info import SparsePauliOp

        op = SparsePauliOp.from_list([("XX", 2), ("YY", 1), ("IZ",2j), ("ZZ",1j)])

        grouped = op.group_commuting()
        grouped_sum = op.group_commuting(qubit_wise=True)

        print(repr(grouped))
        print(repr(grouped_sum))

    .. testoutput::

        [SparsePauliOp(['IZ', 'ZZ'],
                      coeffs=[0.+2.j, 0.+1.j]), SparsePauliOp(['XX', 'YY'],
                      coeffs=[2.+0.j, 1.+0.j])]
        [SparsePauliOp(['XX'],
                      coeffs=[2.+0.j]), SparsePauliOp(['YY'],
                      coeffs=[1.+0.j]), SparsePauliOp(['IZ', 'ZZ'],
                      coeffs=[0.+2.j, 0.+1.j])]

Evolutions
==========
*Back to* `Contents`_

> The :mod:`qiskit.opflow.evolutions` sub-module was created to provide building blocks for Hamiltonian simulation algorithms,

Reviewer 5: ```suggestion
The :mod:`qiskit.opflow.evolutions` submodule was created to provide building blocks for Hamiltonian simulation algorithms,
```
For consistency.

> including various methods for trotterization. The original opflow workflow for hamiltonian simulation did not allow for

Reviewer 5: ```suggestion
including various methods for Trotterization. The original opflow workflow for Hamiltonian simulation did not allow for
```

delayed synthesis of the gates or efficient transpilation of the circuits, so this functionality was migrated to the
``qiskit.synthesis`` :ref:`Evolution <evolution_synthesis>` module.

.. note::

    The :class:`qiskit.opflow.evolutions.PauliTrotterEvolution` class computes evolutions for exponentiated

>     sums of Paulis by converting to the Z basis, rotating with an RZ, changing back, and trotterizing.

Reviewer 5: ```suggestion
    sums of Paulis by converting to the Z basis, rotating with an RZ, changing back, and Trotterizing.
```

    When calling ``.convert()``, the class follows a recursive strategy that involves creating
    :class:`~qiskit.opflow.evolutions.EvolvedOp` placeholders for the operators,
    constructing :class:`.PauliEvolutionGate`\s out of the operator primitives, and supplying one of

>     the desired synthesis methods to perform the trotterization. The methods can be specified via

Reviewer 5: ```suggestion
    the desired synthesis methods to perform the Trotterization. The methods can be specified via
```

    ``string``, which is then inputted into a :class:`~qiskit.opflow.evolutions.TrotterizationFactory`,
    or by supplying a method instance of :class:`qiskit.opflow.evolutions.Trotter`,
    :class:`qiskit.opflow.evolutions.Suzuki` or :class:`qiskit.opflow.evolutions.QDrift`.

>     The different trotterization methods that extend :class:`qiskit.opflow.evolutions.TrotterizationBase` were migrated to

Reviewer 5: ```suggestion
    The different Trotterization methods that extend :class:`qiskit.opflow.evolutions.TrotterizationBase` were migrated to
```

:class:`.Estimator` primitives, send parametrized circuits and observables, and directly receive quasi-probability distributions or
expectation values (respectively). This workflow simplifies considerably the pre-processing and post-processing steps
that previously relied on this module; encouraging us to move away from :mod:`~qiskit.opflow`
and find new paths for developing algorithms based on the :mod:`~qiskit.primitives` interface and
the :mod:`~qiskit.quantum_info` module.

This guide traverses the opflow submodules and provides either a direct alternative
(i.e. using :mod:`~qiskit.quantum_info`), or an explanation of how to replace their functionality in algorithms.

The function equivalency can be roughly summarized as follows:

.. |operator_globals| replace:: ``operator_globals``
.. _operator_globals: https://qiskit.org/documentation/apidoc/opflow.html#operator-globals/

.. list-table::
   :header-rows: 1

   * - Opflow Module
     - Alternative
   * - Operators (:class:`~qiskit.opflow.OperatorBase`, |operator_globals|_ ,
       :mod:`~qiskit.opflow.primitive_ops`,

>        :mod:`~qiskit.opflow.list_ops`\)

Author: No, but you might find a few more scattered around the code, from the time when I thought they were necessary.

.. _qiskit_ibm_runtime.Sampler: https://qiskit.org/documentation/partners/qiskit_ibm_runtime/stubs/qiskit_ibm_runtime.Sampler.html

.. |qiskit_ibm_runtime.Estimator| replace:: ``qiskit_ibm_runtime.Estimator``
.. _qiskit_ibm_runtime.Estimator: https://qiskit.org/documentation/partners/qiskit_ibm_runtime/stubs/qiskit_ibm_runtime.Estimator.html

..  attention::

    Most references to the :class:`qiskit.primitives.Sampler` or :class:`qiskit.primitives.Estimator` in this guide
    can be replaced with instances of the:

    - Aer primitives (|qiskit_aer.primitives.Sampler|_, |qiskit_aer.primitives.Estimator|_)
    - Runtime primitives (|qiskit_ibm_runtime.Sampler|_, |qiskit_ibm_runtime.Estimator|_ )
    - Terra backend primitives (:class:`qiskit.primitives.BackendSampler`, :class:`qiskit.primitives.BackendEstimator`)

    Certain classes, such as the
    :class:`~qiskit.opflow.expectations.AerPauliExpectation`, can only be replaced by a specific primitive instance
    (in this case, |qiskit_aer.primitives.Estimator|_ ), or require a specific option configuration.
    If this is the case, it will be explicitly indicated in the corresponding section.

> Contents

Reviewer 4: I wonder whether we should move this section to above `Background` and below `TD;DR`.

Author: Makes sense

========

This document covers the migration from these opflow sub-modules:

**Operators**

- `Operator Base Class`_
- `Operator Globals`_
- `Primitive and List Ops`_
- `State Functions`_

**Converters**

- `Converters`_
- `Evolutions`_
- `Expectations`_

**Gradients**

- `Gradients`_

Operator Base Class
===================

> *Back to* `Contents`_

Reviewer 4: This is a nice feature. It made me realize that the `TOC` on the right hand side of the theme is not very useful for long documents. I created this issue to see if we can make the TOC follow the scroll so that it's always present: https://github.com/Qiskit/qiskit_sphinx_theme/issues/191

.. |qiskit.quantum_info.BaseOperator| replace:: ``qiskit.quantum_info.BaseOperator``
.. _qiskit.quantum_info.BaseOperator: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/operators/base_operator.py

The :class:`qiskit.opflow.OperatorBase` abstract class can be replaced with |qiskit.quantum_info.BaseOperator|_ ,
keeping in mind that |qiskit.quantum_info.BaseOperator|_ is more generic than its opflow counterpart.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative
   * - :class:`qiskit.opflow.OperatorBase`
     - |qiskit.quantum_info.BaseOperator|_

> 

Reviewer 4: The rendering of list table for the theme is not ideal. The width changes depending on the content. Have you tried using other kinds of tables in Sphinx such as a normal `table`? https://docutils.sourceforge.io/docs/ref/rst/directives.html#toc-entry-21

I can totally see that list table is much easier to right, especially for this case where the entries of each cell is short. Perhaps can we improve the rendering in the qiskit sphinx theme.

<img width="864" alt="Screenshot 2023-03-01 at 12 05 55" src="https://user-images.githubusercontent.com/7631333/222122083-2b4e37b3-ece3-4586-8936-77c7dfb2aa05.png">

..  attention::

    Despite the similar class names, :class:`qiskit.opflow.OperatorBase` and
    |qiskit.quantum_info.BaseOperator|_ are not completely equivalent to each other, and the transition
    should be handled with care. Namely:

    1. :class:`qiskit.opflow.OperatorBase` implements a broader algebra mixin. Some operator overloads that were
    commonly used :mod:`~qiskit.opflow` (for example ``~`` for ``.adjoint()``) are not defined for
    |qiskit.quantum_info.BaseOperator|_. You might want to check the specific
    :mod:`~qiskit.quantum_info` subclass instead.

    2. :class:`qiskit.opflow.OperatorBase` also implements methods such as ``.to_matrix()`` or ``.to_spmatrix()``,
    which are only found in some of the |qiskit.quantum_info.BaseOperator|_ subclasses.

>     See API reference for more information.

Reviewer 4: Which API reference are you referring to here? The reference for individual class? It would be good if you can be a bit more explicit here, provide link here if it is only one or two pages you are referring to.

Operator Globals
================
*Back to* `Contents`_

Opflow provided shortcuts to define common single qubit states, operators, and non-parametrized gates in the
|operator_globals|_ module.

These were mainly used for didactic purposes or quick prototyping, and can easily be replaced by their corresponding
:mod:`~qiskit.quantum_info` class: :class:`~qiskit.quantum_info.Pauli`, :class:`~qiskit.quantum_info.Clifford` or
:class:`~qiskit.quantum_info.Statevector`.

1-Qubit Paulis
--------------
*Back to* `Contents`_

The 1-qubit paulis were commonly used for quick testing of algorithms, as they could be combined to create more complex operators
(for example, ``0.39 * (I ^ Z) + 0.5 * (X ^ X)``).
These operations implicitly created operators of type  :class:`~qiskit.opflow.primitive_ops.PauliSumOp`, and can be replaced by
directly creating a corresponding :class:`~qiskit.quantum_info.SparsePauliOp`, as shown in the examples below.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative
   * - :class:`~qiskit.opflow.X`, :class:`~qiskit.opflow.Y`, :class:`~qiskit.opflow.Z`, :class:`~qiskit.opflow.I`
     - :class:`~qiskit.quantum_info.Pauli`

       ..  tip::

>            For direct compatibility with classes in :mod:`~qiskit.algorithms`, wrap in :class:`~qiskit.quantum_info.SparsePauliOp`.

Reviewer 4: It's a good tip. Can you give an example here. Do you mean something like `X = SparsePauliOp(Pauli('X'))`?

Reviewer: By qiskit.algorithms this really means the new primitive based ones there right. Arguably, like the Estimator they used, they take BaseOperator as type which should mean they (and the Estimator they use) support Pauli directly.

Reviewer: I wonder with these tables similar to this outputs (as per first image below) whether a vertical top align, rather than center, would be better. It would align the two text blocks we most want to compare side by side and leave the tip (or note in other cases) lower down.

![image](https://user-images.githubusercontent.com/40241007/222277452-d0efd99b-7186-4096-a409-bcd99ff88447.png)

So it would look more like this (I edited the above image)

![image](https://user-images.githubusercontent.com/40241007/222277974-826ca30d-5cb4-459a-a458-96be67fcfa2f.png)

Author: @woodsp-ibm I agree that this looks nicer, but I think this is more of a sphinx theme issue in general, right? Maybe better suited for this issue?? https://github.com/Qiskit/qiskit_sphinx_theme/issues/193

Reviewer: I had not looked at the source to see what you used. Normally cell alignment is a choice - maybe what was used gives no choice for the outcome. I made a similar comment as above in that sphinx theme issue you linked.

> .. _1_q_pauli:

Reviewer 4: Should this label be above the section header `1-Qubit-Paulis` in line 166?

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 1: Defining the XX operator</font></a></summary>

>     <br>

Author: Ooh yes!! I love this!

**Opflow**

.. code-block:: python

    >>> from qiskit.opflow import X

>     >>> operator = X ^ X

Reviewer 4: I noticed that [you added](https://github.com/Qiskit/qiskit-terra/pull/9549/commits/fc715320e5143017622353c827073d91fdf2a54d) `>>>` for all code examples with intention to test the code with doctest. We need to do a few things before we can do that. Firstly, you need to change the directive from `code-block` to `doctest` or `testcode`. I personally prefer `testcode` because the code doesn't need to be written as an interactive python session like doctest style with `>>>` which is closer to what most people will use qiskit (that is in Jupyter notebook / python script). 

Secondly, doctest doesn't not test the code during docs build using the default building command (`sphinx-build -b`). It needs to run with a different commend (`sphinx-build -b doctest`, which only does testing, not building) to do doctest which Terra doesn't have currently. This is implemented in the sphinx theme for building the docs guide (see: https://github.com/Qiskit/qiskit_sphinx_theme/blob/2b5ced2061f842d3ac3424080d014cf028eb747a/tox.ini#L21-L27 and https://github.com/Qiskit/qiskit_sphinx_theme/blob/2b5ced2061f842d3ac3424080d014cf028eb747a/.github/workflows/docs-guide-deploy.yml#L25-L28) We plan to implement that in the future but it's out of the scope of this PR and it should not be blocked by that. 

See https://qiskit.github.io/qiskit_sphinx_theme/sphinx_guide/how_to_add_code.html#run-the-tests for more details.

That said, I actually think we shouldn't test the code in migration guides, because these code will be deprecated and will raise errors when they are removed if we test the code. And since we are not testing the code, I prefer not having `>>>` each code line. Sorry it might be a bit trouble for you to revert it back.

Reviewer: It is possible to add skip directives, so its testable while the code is there - of course it would mean editing it once things are removed. For me it would be preferable to have this tested by CI to ensure it works and continues to do so as described until the code is removed.

Reviewer 4: I agree. Thinking a bit more about it, we will anyway need to edit the migration guides after the features are removed because cross reference like `:class:`, `:meth:` will break.

    >>> state1
    DictStateFn({'01': 1}, coeff=1.0, is_measurement=False)

    >>> print(state2)
    CircuitStateFn(
         ┌───┐┌───┐
    q_0: ┤ X ├┤ H ├
         ├───┤└───┘
    q_1: ┤ H ├─────
         └───┘
    )

**Alternative**

.. code-block:: python

    >>> from qiskit import QuantumCircuit
    >>> from qiskit.quantum_info import StabilizerState, Statevector

    >>> qc_zero = QuantumCircuit(1)
    >>> qc_one = qc_zero.copy()
    >>> qc_one.x(0)
    >>> state1 = Statevector(qc_zero) ^ Statevector(qc_one)

    >>> qc_plus = qc_zero.copy()
    >>> qc_plus.h(0)
    >>> qc_minus = qc_one.copy()
    >>> qc_minus.h(0)
    >>> state2 = StabilizerState(qc_plus) ^ StabilizerState(qc_minus)

    >>> state1
    Statevector([0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
                dims=(2, 2))

    >>> state2
    StabilizerState(StabilizerTable: ['-IX', '+XI'])

.. raw:: html

   </details>

Primitive and List Ops
======================
*Back to* `Contents`_

Most of the workflows that previously relied in components from :mod:`~qiskit.opflow.primitive_ops` and
:mod:`~qiskit.opflow.list_ops` can now leverage elements from :mod:`~qiskit.quantum_info`\'s
operators instead.

> Some of these classes do not require a 1-1 replacement because they were created to interface with other

Reviewer: I would lean somewhat towards removing this last sentence as each of the classes are covered below, and since its rather vague as to what the classes in question might be. I think you cover all the classes in this category that comprise the public API - if its internal to opflow then at this level it should not be a consideration right.

    >>> qubit_op
    SparsePauliOp(['XYZY'],
                  coeffs=[0.-6.j])

.. raw:: html

   </details>

.. _z2_sym:

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 2: <code>Z2Symmetries</code> and <code>TaperedPauliSumOp</code></font></a></summary>
    <br>

**Opflow**

.. code-block:: python

    >>> from qiskit.opflow import PauliSumOp, Z2Symmetries, TaperedPauliSumOp

    >>> qubit_op = PauliSumOp.from_list(
    ...    [
    ...    ("II", -1.0537076071291125),
    ...    ("IZ", 0.393983679438514),
    ...    ("ZI", -0.39398367943851387),
    ...    ("ZZ", -0.01123658523318205),
    ...    ("XX", 0.1812888082114961),
    ...    ]
    ... )
    >>> z2_symmetries = Z2Symmetries.find_Z2_symmetries(qubit_op)
    >>> tapered_op = z2_symmetries.taper(qubit_op)

    # can be represented as:
    >>> tapered_op = TaperedPauliSumOp(qubit_op.primitive, z2_symmetries)

    >>> print(z2_symmetries)
    Z2 symmetries:
    Symmetries:
    ZZ
    Single-Qubit Pauli X:
    IX
    Cliffords:
    0.7071067811865475 * ZZ
    + 0.7071067811865475 * IX
    Qubit index:
    [0]
    Tapering values:
      - Possible values: [1], [-1]

>     >>> tapered_op

Reviewer: Above tapered_op is redefined. What were you wanting to print here? This is the original operator - versus below where its a list of the operators according to the taperings.

    >>> hamiltonian = MatrixOp([[0, 1], [1, 0]])
    >>> evolution = MatrixEvolution()
    >>> evol_result = evolution.convert(hamiltonian.exp_i())
    >>> evolved_state = evol_result.to_circuit()

    >>> print(evolved_state.decompose().decompose())
       ┌────────────────┐
    q: ┤ U3(2,-π/2,π/2) ├
       └────────────────┘

**Alternative**

.. code-block:: python

    >>> from qiskit.quantum_info import SparsePauliOp
    >>> from qiskit.extensions import HamiltonianGate
    >>> from qiskit import QuantumCircuit

    >>> evol_gate = HamiltonianGate([[0, 1], [1, 0]], 1)
    >>> evolved_state = QuantumCircuit(1)
    >>> evolved_state.append(evol_gate, [0])

    >>> print(evolved_state.decompose().decompose())
       ┌────────────────┐
    q: ┤ U3(2,-π/2,π/2) ├
       └────────────────┘

.. raw:: html

    </details>

Expectations
============
*Back to* `Contents`_

Expectations are converters which enable the computation of the expectation value of an observable with respect to some state function.
This functionality can now be found in the Estimator primitive.

Algorithm-Agnostic Expectations
-------------------------------
*Back to* `Contents`_

> 

Reviewer 4: Maybe here is a good place to remind people that there are different implementations of primitives as noted `attention` box in the line 82. Maybe link to that.

.. list-table::
   :header-rows: 1

   * - Opflow
     - Alternative

   * - :class:`~qiskit.opflow.expectations.ExpectationFactory`
     - No direct replacement. This class was used to create instances of one of the classes listed below.

   * - :class:`~qiskit.opflow.expectations.AerPauliExpectation`
     - Use ``Estimator`` primitive from |qiskit_aer.primitives|_ with ``approximation=True`` and ``shots=None`` as ``run_options``.
       See example below.

   * - :class:`~qiskit.opflow.expectations.MatrixExpectation`
     - Use :class:`~qiskit.primitives.Estimator` primitive from :mod:`qiskit` (if no shots are set, it performs an exact Statevector calculation).
       See example below.

   * - :class:`~qiskit.opflow.expectations.PauliExpectation`
     - Use any Estimator primitive (for :class:`qiskit.primitives.Estimator`, set ``shots!=None`` for a shot-based
       simulation, for |qiskit_aer.primitives.Estimator|_ , this is the default).

.. _expect_state:

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 1: Aer Pauli expectation</font></a></summary>
    <br>

**Opflow**

.. code-block:: python

>     >>> from qiskit.opflow import X, Minus, StateFn, AerPauliExpectation, CircuitSampler
>     >>> from qiskit.utils import QuantumInstance
>     >>> from qiskit_aer import Aer
> 
>     >>> backend = Aer.get_backend("aer_simulator")

Reviewer 4: ```suggestion
    >>> from qiskit.opflow import X, Minus, StateFn, AerPauliExpectation, CircuitSampler
    >>> from qiskit.utils import QuantumInstance
    >>> from qiskit_aer import AerSimulator

    >>> backend = AerSimulator()
```

    >>> q_instance = QuantumInstance(backend)

    >>> sampler = CircuitSampler(q_instance, attach_results=True)
    >>> expectation = AerPauliExpectation()

    >>> state = Minus
    >>> operator = 1j * X

    >>> converted_meas = expectation.convert(StateFn(operator, is_measurement=True) @ state)
    >>> expectation_value = sampler.convert(converted_meas).eval()

    >>> print(expectation_value)
    -1j

**Alternative**

.. code-block:: python

    >>> from qiskit.quantum_info import SparsePauliOp
    >>> from qiskit import QuantumCircuit

>     >>> from qiskit_aer.primitives import Estimator as AerEstimator

Reviewer 4: ```suggestion
    >>> from qiskit_aer.primitives import Estimator
```

Do we really need import `Estimator as AerEstimator`? In this example, there is no clash of Aer's Estimator with Terra's Estimator. I think one great benefits of using the same class name is that it can be changed among different implementations of primitives easily.

    >>> estimator = AerEstimator(run_options={"approximation": True, "shots": None})

    >>> op = SparsePauliOp.from_list([("X", 1j)])
    >>> states_op = QuantumCircuit(1)
    >>> states_op.x(0)
    >>> states_op.h(0)

    >>> expectation_value = estimator.run(states_op, op).result().values

    >>> print(expectation_value)
    [0.-1.j]

.. raw:: html

    </details>

.. _matrix_state:

.. raw:: html

    <details>
    <summary><a><font size="+1">Example 2: Matrix expectation</font></a></summary>
    <br>

**Opflow**

.. code-block:: python

>     >>> from qiskit_aer import Aer
>     >>> from qiskit.opflow import X, H, I, MatrixExpectation, ListOp, StateFn
>     >>> from qiskit.utils import QuantumInstance
> 
>     >>> backend = Aer.get_backend("statevector_simulator")

Reviewer 4: ```suggestion
    >>> from qiskit.opflow import X, H, I, MatrixExpectation, ListOp, StateFn
    >>> from qiskit.utils import QuantumInstance
    >>> from qiskit_aer import AerSimulator

    >>> backend = AerSimulator(method='statevector')
```

            from qiskit.quantum_info import Pauli
            X = Pauli('X')
            op = X ^ X

     -

        .. code-block:: python

            from qiskit.quantum_info import Pauli, SparsePauliOp
            op = Pauli('X') ^ Pauli('X') 
            
            # equivalent to:
            op = SparsePauliOp('XX')

Common non-parametrized gates (Clifford)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. list-table:: Migration of ``qiskit.opflow.operator_globals`` (2/3)
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * - :class:`~qiskit.opflow.CX`, :class:`~qiskit.opflow.S`, :class:`~qiskit.opflow.H`, :class:`~qiskit.opflow.T`, :class:`~qiskit.opflow.CZ`, :class:`~qiskit.opflow.Swap`
     - Append corresponding gate to :class:`~qiskit.QuantumCircuit` + :class:`~qiskit.quantum_info.Clifford` + ``.to_operator()``
     -

   * -

        .. code-block:: python

            from qiskit.opflow import H
            op = H ^ H

     -

        .. code-block:: python

            from qiskit import QuantumCircuit
            from qiskit.quantum_info import Clifford
            qc = QuantumCircuit(2)
            qc.h(0)
            qc.h(1)

>             op = Clifford(qc).to_operator()

Author: Is there any "good" alternative you can think of? Or should I just mention that it is possible to do the `to_operator` but not really recommendable?

>             op = Clifford(qc).to_operator()

Reviewer 6: I think we should keep Clifford for now. Users should be able to convert it to the form they want when they need it.

            # or...
            qc = QuantumCircuit(1)
            qc.h(0)
            H = Clifford(qc).to_operator()
            op = H ^ H

     -

1-Qubit States
~~~~~~~~~~~~~~
.. list-table:: Migration of ``qiskit.opflow.operator_globals`` (3/3)
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * - :class:`~qiskit.opflow.Zero`, :class:`~qiskit.opflow.One`, :class:`~qiskit.opflow.Plus`, :class:`~qiskit.opflow.Minus`
     - :class:`~qiskit.quantum_info.Statevector` or :class:`~qiskit.QuantumCircuit` directly
     -

   * -

        .. code-block:: python

            from qiskit.opflow import Zero, One, Plus, Minus

            state1 = Zero ^ One
            state2 = Plus ^ Minus

     -

        .. code-block:: python

            from qiskit import QuantumCircuit
            from qiskit.quantum_info import Statevector

            qc_zero = QuantumCircuit(1)
            qc_one = copy(qc_zero)
            qc_one.x(0)

>             state1 = Statevector(qc_zero) ^ Statevector(qc_one)

Reviewer 6: StabilizerState is efficient. (It is not exactly the same with opflow, so it depends on the purpose.)

     -
   * - :class:`~qiskit.opflow.QDrift`
     - :class:`~synthesis.QDrift`
     -

.. list-table:: Migration of ``qiskit.opflow.evolutions.evolutions``
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * - :class:`~qiskit.opflow.EvolutionFactory`
     - This class is no longer necessary.
     -
   * - :class:`~qiskit.opflow.EvolvedOp`
     - :class:`~synthesis.SuzukiTrotter`
     - This class is no longer necessary
   * - :class:`~qiskit.opflow.MatrixEvolution`
     - :class:`~HamiltonianGate`
     -
   * - :class:`~qiskit.opflow.PauliTrotterEvolution`
     - :class:`~PauliEvolutionGate`
     -

Expectations
------------
Expectations are converters which enable the computation of the expectation value of an observable with respect to some state function.
This functionality can now be found in the estimator primitive.

Algorithm-Agnostic Expectations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. list-table:: Migration of ``qiskit.opflow.expectations``
   :header-rows: 1

   * - opflow
     - alternative
     - notes
   * - :class:`~qiskit.opflow.ExpectationFactory`
     - No replacement needed.
     -
   * - :class:`~qiskit.opflow.AerPauliExpectation`

>      - Use :class:`~Estimator` primitive from ``qiskit_aer`` instead.

Reviewer 6: `approximation=True` option corresponds to `AerPauliExpectation`. If `shots=None` it return exact values.

Author: Got it, thanks!

computations were left to the user to manage. Opflow helped bridge that gap, it allowed to wrap circuits and
observables into operator classes that could be algebraically manipulated, so that the final result's expectation
values could be easily computed following different methods.

This basic opflow functionality is covered by  its core submodules: the ``operators`` submodule
(including :mod:`~qiskit.opflow.operator_globals`, :mod:`~qiskit.opflow.list_ops`, :mod:`~qiskit.opflow.primitive_ops`, and :mod:`~qiskit.opflow.state_fns`),
the :mod:`~qiskit.opflow.converters` submodule, and the :mod:`~qiskit.opflow.expectations` submodule.
Following this reference framework of ``operators``, :mod:`~qiskit.opflow.converters` and :mod:`~qiskit.opflow.expectations`, opflow includes more
algorithm-specific functionality, which can be found in the :mod:`~qiskit.opflow.evolutions` submodule (specific for hamiltonian
simulation algorithms), as well as the :mod:`~qiskit.opflow.gradients` submodule (applied in multiple machine learning and optimization
use-cases). Some classes from the core modules mentioned above are also algorithm or application-specific,

> for example the :obj:`~CVarMeasurement` or the :obj:`~Z2Symmetries`.

Author: Yes, I will remove both this paragraph and the previous one from the guide. We are planning to write a blog post diving deeper into why is opflow being deprecated, and this kind of explanation of the background fits better over there anyways.

   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * - :class:`~qiskit.opflow.Zero`, :class:`~qiskit.opflow.One`, :class:`~qiskit.opflow.Plus`, :class:`~qiskit.opflow.Minus`
     - :class:`~qiskit.quantum_info.StabilizerState` or :class:`~qiskit.quantum_info.Statevector` or :class:`~qiskit.QuantumCircuit`, depending on the use case
     - In principle, :class:`~qiskit.quantum_info.StabilizerState` is the most efficient replacement for :class:`~qiskit.opflow` states, but the functionality is not identical. See API ref. for more info.

   * -

        .. code-block:: python

            from qiskit.opflow import Zero, One, Plus, Minus

            state1 = Zero ^ One
            state2 = Plus ^ Minus

     -

        .. code-block:: python

            from qiskit import QuantumCircuit
            from qiskit.quantum_info import StabilizerState

            qc_zero = QuantumCircuit(1)
            qc_one = copy(qc_zero)
            qc_one.x(0)
            state1 = StabilizerState(qc_zero) ^ StabilizerState(qc_one)

            qc_plus = copy(qc_zero)
            qc_plus.h(0)
            qc_minus = copy(qc_one)
            qc_minus.h(0)

>             state2 = StabilizerState(qc_plus) ^ StabilizerState(qc_minus)

Author: I know, but I don't know what better format we could use here. It looks like the tables don't adapt to the screen size too, if they did, these examples would probably not wrap.

>             state2 = StabilizerState(qc_plus) ^ StabilizerState(qc_minus)

Author: I have decided to merge the "alternative" and ''notes" columns, as the third one was almost empty. This has improved a bit the look of the tables. But still, I think there are probably other ways in which this could be improved.

>             state2 = StabilizerState(qc_plus) ^ StabilizerState(qc_minus)

Reviewer: In other guides we have more done things sequentially as below. Here one gets a full width for the code. The side-by-side for more compact examples seems nicer, but if the code starts wrapping it rather detracts from that for me.

**Previously**
```
old way example code
```

**New**
```
new way example code
```

     -

Primitive and List Ops
----------------------
Most of the workflows that previously relied in components from :mod:`~qiskit.opflow.primitive_ops` and :mod:`~qiskit.opflow.list_ops` can now
leverage elements from :mod:`~qiskit.quantum_info.operators` instead. Some of these classes do not require a 1-1 replacement because
they were created to interface with other opflow components.

PrimitiveOps
~~~~~~~~~~~~~~

.. list-table:: Migration of ``qiskit.opflow.primitive_ops``
   :header-rows: 1

   * - opflow
     - alternative
     - notes

>    * - :class:`~qiskit.opflow.PrimitiveOp`

Author: I had noticed too, and will update the links.

     - Can directly use :class:`~qiskit.quantum_info.Operator``
   * - :class:`~qiskit.opflow.CircuitOp`
     - No direct replacement
     - Can directly use :class:`~qiskit.QuantumCircuit`
   * - :class:`~qiskit.opflow.MatrixOp`
     - :class:`~qiskit.quantum_info.Operator``
     -
   * - :class:`~qiskit.opflow.PauliOp`
     - :class:`~qiskit.quantum_info.Pauli`
     - For direct compatibility with classes in :mod:`~qiskit.algorithms`, wrap in :class:`~qiskit.quantum_info.SparsePauliOp`
   * - :class:`~qiskit.opflow.PauliSumOp`
     - :class:`~qiskit.quantum_info.SparsePauliOp`
     - See example below
   * - :class:`~qiskit.opflow.TaperedPauliSumOp`
     - This class was used to combine a :class:`~PauliSumOp` with its identified symmetries in one object. It has been deprecated without replacement
     - See ``Z2Symmetries`` example for updated workflow
   * - :class:`~qiskit.opflow.Z2Symmetries`
     - :class:`~qiskit.quantum_info.Z2Symmetries`
     - See example below

PrimitiveOps Examples
~~~~~~~~~~~~~~~~~~~~~
.. list-table::
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * -  ``PauliSumOp`` **Example:**

        .. code-block:: python

>             from qiskit.opflow import PuliSumOp

Author: @mtreinish?

>             from qiskit.opflow import PuliSumOp

Author: So @woodsp-ibm, I managed to find an inbetween solution, I formatted the code examples to be runnable locally with pytest's `doctest`. I don't know if I like having all of the `>>>` in the code, but at least we can make sure that it runs.

>             from qiskit.opflow import PuliSumOp

Reviewer: It is also possible to use ..testcode:: and ..testoutput:: directives (they are supported by the sphinx.ext.doctest, which we use in the apps and imagine would be used here too). By default the output puts these in separate highlighted blocks (though you can hide the testoutput if desired) but there is no need for >>> when using these.

I know though we used >>> in Nature etc since formatting with these former seemed harder to get things to come out right/as wanted etc. The >>> was always more concise and seemed easier to add and have it come out as expected in among other text. Maybe there is styling that can applied to testcode/output so things can be improved, not sure.

E.g.

```
.. testcode::

    x = 1 + 3
    print(x)

.. testoutput::

    4

```
comes out as
![image](https://user-images.githubusercontent.com/40241007/222256752-15e0a92f-04b8-46b9-9e2f-520cc2fc6ae7.png)

                ("ZI", -0.39398367943851387),
                ("ZZ", -0.01123658523318205),
                ("XX", 0.1812888082114961),
                ]
            )
            z2_symmetries = Z2Symmetries.find_Z2_symmetries(qubit_op)
            tapered_op = z2_symmetries.taper(qubit_op)
            # can be represented as:
            tapered_op = TaperedPauliSumOp(primitive, z2_symmetries)
     -

        .. code-block:: python

            from qiskit.quantum_info import SparsePauliOp, Z2Symmetries

            qubit_op = SparsePauliOp.from_list(
                [
                    ("II", -1.0537076071291125),
                    ("IZ", 0.393983679438514),
                    ("ZI", -0.39398367943851387),
                    ("ZZ", -0.01123658523318205),
                    ("XX", 0.1812888082114961),
                ]
            )
            z2_symmetries = Z2Symmetries.find_z2_symmetries(qubit_op)
            tapered_op = z2_symmetries.taper(qubit_op)
     -

ListOps
~~~~~~~
.. list-table:: Migration of ``qiskit.opflow.list_ops``
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * - :class:`~qiskit.opflow.ListOp`

>      - No direct replacement. This class was used internally within opflow.

Reviewer: I would delete this bit for many cases here - `This class was used internally within opflow.` leaving just the fact that there is no direct replacement. These classes are exposed public classes for end users. Sometimes they did get built implicitly through the opflow expressions, but you could do it explicilitly - this https://qiskit.org/documentation/tutorials/operators/01_operator_flow.html for instance builds out ListOps and with expressions a ComposedOp etc.

Now we can compose and tensor operators. opflow was designed for lazy evaluation so its a little different but its similar. The SummedOp, if you have a sum of Paulis, which is what it was used for to start with mostly, before we did the more efficient PauliSumOp (based on SparsePauliOp) - then a reasonable alternative for those common cases is the SparsePauliOp. More generally the SummedOp could contain any other opflow operator so it general there is no alternative.

            from qiskit.primitives import Estimator
            from qiskit.quantum_info import SparsePauliOp

            state = QuantumCircuit(1)
            state.h(0)
            hamiltonian = SparsePauliOp.from_list([('X', 1), ('Z',1)])

            estimator = Estimator()
            expectation_value = estimator.run(state, hamiltonian).result().values

     -

Two Qubit Reduction
~~~~~~~~~~~~~~~~~~~~
.. list-table:: Migration of ``qiskit.opflow.TwoQubitReduction``
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * -  :class:`~qiskit.opflow.TwoQubitReduction`
     -  No direct replacement.
     -  This class implements a chemistry-specific reduction for the ``ParityMapper`` class in ``qiskit-nature``.
        The general symmetry logic this mapper depends on has been refactored to other classes in :mod:`~qiskit.quantum_info`,
        so this specific :mod:`~qiskit.opflow` implementation is no longer necessary.

Other Converters
~~~~~~~~~~~~~~~~~

.. list-table:: Migration of ``qiskit.opflow.converters``
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * - :class:`~qiskit.opflow.AbelianGrouper`
     - No direct replacement.

>      - This class allowed a sum a of Pauli operators to be grouped. These type of groupings are now left to the primitives to handle (??).

Reviewer: > These type of groupings are now left to the primitives to handle (??).

Yes, the given an operator which is a sum of Paulis then this is now the responsibilty of an Estimator - Aer Estimator has just added the support - the Runtime Estimator I believe does this as does the BackendEstimator (for info, without it a circuit is run for each Pauli, with it one for each group which can be way less circuits run so better performance)

     -
   * - :class:`~qiskit.opflow.QDrift`
     - :class:`~synthesis.QDrift`
     -

.. list-table:: Migration of ``qiskit.opflow.evolutions.evolutions``
   :header-rows: 1

   * - opflow
     - alternative
     - notes

   * - :class:`~qiskit.opflow.EvolutionFactory`
     - This class is no longer necessary.
     -
   * - :class:`~qiskit.opflow.EvolvedOp`
     - :class:`~synthesis.SuzukiTrotter`
     - This class is no longer necessary
   * - :class:`~qiskit.opflow.MatrixEvolution`
     - :class:`~HamiltonianGate`
     -
   * - :class:`~qiskit.opflow.PauliTrotterEvolution`
     - :class:`~PauliEvolutionGate`
     -

Expectations
------------
Expectations are converters which enable the computation of the expectation value of an observable with respect to some state function.
This functionality can now be found in the estimator primitive.

Algorithm-Agnostic Expectations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. list-table:: Migration of ``qiskit.opflow.expectations``
   :header-rows: 1

   * - opflow
     - alternative
     - notes
   * - :class:`~qiskit.opflow.ExpectationFactory`
     - No direct replacement.
     -
   * - :class:`~qiskit.opflow.AerPauliExpectation`

>      - Use :class:`~Estimator` primitive from ``qiskit_aer`` with ``approximation=True``

Reviewer: And `shots=None` - the AerPauliExpectation gave an exact answer using Aer's special expectation instruction I think it needs both shots=None and approximation=True for the same net outcome (though its confusing to me that having something called approximation True gives an Exact outcome!!)
